Added ability to set alternate keys in fluent syntax

nsarris opened the pull request:
Fix for issue #421 

Added an alternative to Key() named AltKey() that can be used with both an anonymous object or a dictionary (names are mandatory)

Followed the original pattern where non matching keys are ignored

Added an associated test

watfordgnf commented on src/Simple.OData.Client.Core/Fluent/IFluentClient.cs:
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
         /// <returns>Self.</returns>
         FT Key(T entryKey);
 
+        /// Sets the specified alternate entry key.
```
Missing a `<summary>` tag.

nsarris replied:
Fixed

watfordgnf commented on src/Simple.OData.Client.Core/Fluent/IFluentClient.cs:
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
+        /// <param name="entryKey">The entry key.</param>
+        /// <returns>Self.</returns>
+        FT AltKey(object entryKey);
+        /// <summary>
```
```suggestion

        /// <summary>
```

nsarris replied:
I noticed that overloads are grouped together without spaces so I kept it like that. Let me know if you still want that change

watfordgnf replied:
No that's okay if that's the original way.

watfordgnf commented on src/Simple.OData.Client.Core/Fluent/IFluentClient.cs:
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
 
+        /// Sets the specified alternate entry key.
+        /// </summary>
+        /// <param name="entryKey">The entry key.</param>
```
```suggestion
        /// <param name="entryKey">The alternate entry key.</param>
```

nsarris replied:
Fixed

watfordgnf commented on src/Simple.OData.Client.Core/Fluent/IFluentClient.cs:
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
+        /// <summary>
+        /// Sets the specified alternate entry key.
+        /// </summary>
+        /// <param name="entryKey">The entry key.</param>
```
```suggestion
        /// <param name="entryKey">The alternate entry key.</param>
```

nsarris replied:
Fixed

watfordgnf commented on src/Simple.OData.Client.Core/Fluent/FluentCommand.cs:
```diff
@@ -258,6 +260,31 @@ public FluentCommand Key(IDictionary<string, object> key)
+            if (key != null && TypeCache.IsAnonymousType(key.GetType()))
+            {
+                _details.NamedKeyValues = TypeCache.ToDictionary(key);
+            }
```
This should probably raise an argument exception if an object is used that cannot be made into a `NamedKeyValues` dictionary. As this stands it would also not support `client.AltKey("name")` looking up that there is only a single alternate key; an exception should be raised in that instance too.

nsarris replied:
That was my initial thought but reading through the code (for that existing Key() implementation) I noticed that
- No validation is performed when setting the values using Key()
- Validation is only performed when building the CommandText at which point the only thing done is filtering out invalid names (if named pairs given) or zipping together the values (if values given)
- Alternate Keys are not handled in the metadata

So I followed that pattern. If you want I can perform another PR that handles Alternate Key meta and performs actual validation on both Primary or Alternate keys.  For alternate keys I could also make the discovery of FilterAsKey working for alternate keys.

watfordgnf replied:
If there is not any validation along the `Key` path, I'm okay with this as-is for this PR. Certainly a subsequent PR could include the validation.

nsarris replied:
In that case, now that I think about it a separate AltKey method would be obsolete. OData it self has one syntax for primary and alternate key so let's follow that. Let me follow up on this PR and consume the alternate key metadata to have the Key method handle everything.

watfordgnf commented on src/Simple.OData.Client.Core/Fluent/FluentCommand.cs:
```diff
@@ -579,30 +606,25 @@ private string Format()
-                var namedKeyValues = new Dictionary<string, object>();
-                for (var index = 0; index < keyNames.Count; index++)
+
+                return (_details.NamedKeyValues?.Select(x => new
```
Out of curiosity, why was this changed from the original implementation?

nsarris replied:
The original implementation was:
- When named values given, filter out invalid names using the primary key definition
- When values given, zip together with the primary key names

I tried to perform minimal changes so I tried to squeeze alternate keys in there too. The intent was to discover matching model fields so that I only keep the original behaviour for primary and not alternate keys. Looking at it again I forgot that filter so I commited a fix and relevant tests for that, replicating the original behaviour (filter out invalid names when primary key). If we perform the validation discussed on the previous review item then this can go.

watfordgnf commented on src/Simple.OData.Client.UnitTests/FluentApi/FindTypedTests.cs:
```diff
@@ -287,6 +287,17 @@ public async Task Get()
         }
 
+        [Fact]
+        public async Task GetWithAltKey()
```
A test which uses the `IDictionary<string, object>` overload should be included as well.

nsarris replied:
I refactored the object overload to call the dictionary overload underneath so the test covers both cases. Let me know if still want a seperate test.

watfordgnf commented on src/Simple.OData.Client.Core/Fluent/CommandDetails.cs:
```diff
@@ -14,6 +14,7 @@ class CommandDetails
         public string DynamicPropertiesContainerName { get; set; }
         public string FunctionName { get; set; }
         public string ActionName { get; set; }
+        public bool IsAlternateKey { get; set; }
```
A comment pointing to the current community spec would be helpful for future editors.

nsarris replied:
Please elaborate, I'm not sure what I have to do here :)

watfordgnf replied:
I was unfamiliar with the alternate keys syntax. Basically adding a comment with a pointer to the spec that governs them would be helpful: https://github.com/OData/vocabularies/blob/master/OData.Community.Keys.V1.md

nsarris replied:
Alright, want that comment right there on the IsAlternateKey property?

watfordgnf replied:
After reviewing your updated changes I think this would be best in a the V4 Metadata code by the `GetAlternateKeyPropertyNames`.

nsarris replied:
Added summary comments to interface declaration and both V3/V4 implementations of GetAlternateKeyPropertyNames. It would be nice if this were done for all the metadata methods.

nsarris commented:
Refactored the code to use the Key() syntax (removed the AltKey syntax altogether). This mimics the behaviour of the odata spec on keys and alternate keys, while being backwards compatible with Simple.

- A collection of values is treated as a primary key by zipping the values together with the names. If less values are provided the query will fail on execution. If more are provided the extra values are ignored.
- A key/value pair collection scans all keys (primary first) for exact matches and selects the first occurance. If no match is found it falls back to trying to extract the primary key from the key/value pairs. This is necessary for backwards compatibility (this is how a whole entity passed in Key is treated).

We can handle the entity instance as key seperately (by making sure the type or type name match before treating it as a key). Then it would be possible to actually throw an exception when no other matches are found. I suggest we talk about this in another PR.

I added a collection of tests for any combination of keys I could think about.

watfordgnf commented:
Agreed on using another PR to tackle some of the validation logic (one I considered was calling `Key` without any primary key in the case where no alternate keys exist).

watfordgnf commented on src/Simple.OData.Client.V4.Adapter/Metadata.cs:
```diff
@@ -229,10 +229,24 @@ public override IEnumerable<string> GetDeclaredKeyPropertyNames(string collectio
 
         public override string GetFunctionFullName(string functionName)
         {
+            GetAlternateKeyPropertyNames("");
```
Was this testing code, or is there a side effect here that is important?

nsarris replied:
No idea how that got there, possible an accidental paste. Removed

object commented:
Thanks for the PR. Tested and merged.
